File: benchmarks/fixtures/webapp_csharp/Services/BaseService.cs
namespace Webapp.Services;

using Webapp.Util;

/// <summary>
/// Common base for all application services.
/// </summary>
public abstract class BaseService
{
    private static readonly Logger Log = Logger.GetLogger("services.base");
    private readonly string _serviceName;
    private bool _initialized;

    protected BaseService(string serviceName)
    {
        _serviceName = serviceName;
        _initialized = false;
    }

    public string Name => _serviceName;

    public void Initialize()
    {
        Log.Info($"Initializing service: {_serviceName}");
        _initialized = true;
    }

    protected void RequireInitialized()
    {
        if (!_initialized)
        {
            Log.Warn($"{_serviceName} is not initialized");
        }
    }
}

File: benchmarks/fixtures/webapp_cpp/services/base_service.cpp
#include "services/base_service.h"

#include <string>
#include <utility>

#include "util/logger.h"

namespace webapp {
namespace services {

BaseService::BaseService(std::string service_name)
    : log_(util::Logger::get_logger("services.base")),
      service_name_(std::move(service_name)),
      initialized_(false) {}

/// Mark the service ready to serve traffic.
void BaseService::initialize() {
    log_.info("Initializing service: " + service_name_);
    initialized_ = true;
}

const std::string& BaseService::name() const {
    return service_name_;
}

bool BaseService::initialized() const {
    return initialized_;
}

/// Warn when a caller reached the service before initialize().
void BaseService::require_initialized() const {
    if (!initialized_) {
        log_.warn(service_name_ + " is not initialized");
    }
}

}  // namespace services
}  // namespace webapp
